feat(order/category/DecLinOrder): updates to concrete_category, and instances for order categories

kim-em opened the pull request:
This is stacked on top of @urkud's #1380, refactoring concrete categories.

The only additional commit beyond that is https://github.com/leanprover-community/mathlib/commit/010709787463a7615ebb33da935f69cff96beb52.

I wanted to test that refactor out, by adding the categories `Preorder`, `PartialOrder`, `LinOrder`, `DecLinOrder`, as an experiment with a (slightly) deep algebraic hierarchy.

I did end up making one substantive change to Yury's work on concrete categories, in particular there are now two main sources of `has_coe_to_sort` instances. As well as  
```
instance : has_coe_to_sort (bundled c) :=	
{ S := Type u, coe := bundled.α }	
```
we have
```
instance induced_category.has_coe_to_sort {C D : Type (u+1)} [concrete_category D] (f : C → D) :
  has_coe_to_sort (induced_category D f) :=
concrete_category.has_coe_to_sort (induced_category D f)
```
Using this, I've switch to defining many concrete categories directly via `induced_category`, rather than providing additional `bundled_hom` instances. (A small change here -- I made the "origin category" argument of `induced_category` explicit, as it's very often useful to be able to give it by hand.)

There is also the more general
```
def concrete_category.has_coe_to_sort (C : Type u) [concrete_category C] : has_coe_to_sort C :=
{ S := Sort u, coe := (concrete_category.forget C).obj }
```
which would work too, but because it applies to any type is too expensive to have as a global instance.

Defining `DecLinOrder` worked very smoothly! I've also tweaked all the constructions of concrete categories.

kim-em committed:
trying to use bundled homs

kim-em committed:
bundled monoids should use bundled homs

kim-em committed:
fixes

kim-em committed:
Merge branch 'master' into Mon_bundled_hom

kim-em committed:
fixes

urkud committed:
refactor(*): rewrite concrete categories

urkud committed:
Add module documentation

kim-em committed:
local instance

urkud committed:
Merge branch 'Mon_bundled_hom' into concrete-categories

Compile fails. I'll resolve actual conflicts later.

urkud committed:
Move files around; don't touch content yet

urkud committed:
Move code around, fix some compile errors

urkud committed:
Add `unbundled_hom.lean`

urkud committed:
Merge branch 'master' into concrete-categories

urkud committed:
Turn `hom` into an argument of `(un)bundled_hom`

For some reason, it helps Lean find coercions from `Ring` category
morphisms to functions.

urkud committed:
Define `unbundled_hom.mk_has_forget`; fix compile

urkud committed:
Merge branch 'master' into concrete-categories

urkud committed:
Add some documentation

urkud committed:
Fix compile

kim-em committed:
chore(category_theory): require morphisms live in Type

kim-em committed:
move back to Type

kim-em committed:
tweaks to doc-comments

kim-em committed:
fixing some formatting

urkud committed:
Revert most of the changes to `data/mv_polynomials`

urkud committed:
Merge branch 'master' into concrete-categories

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

urkud committed:
Drop unused argument, add a comment

urkud committed:
Merge branch 'master' into concrete-categories

kim-em committed:
merge

urkud committed:
Merge branch 'master' into concrete-categories

kim-em committed:
Merge remote-tracking branch 'origin/master' into category_no_sorts

urkud committed:
Simplify universe levels in `concrete_category/basic`.

Still fails to compile.

urkud committed:
Simplify universe levels in `concrete_category/{un,}bundled_hom`

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

kim-em committed:
fixes

urkud committed:
Fix an `import`

urkud committed:
Merge branch 'master' into concrete-categories

urkud committed:
Doc cleanup

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

urkud committed:
Merge branch 'master' into concrete-categories

kim-em committed:
feat(order/category/DecLinOrder): updates to concrete_category, and instances of order categories.

kim-em committed:
fix

kim-em committed:
fix

kim-em committed:
fix

kim-em committed:
fixes

kim-em committed:
...

kim-em committed:
...

kim-em committed:
Merge branch 'category_no_sorts' into concrete-categories

kim-em committed:
update post #1412

kim-em committed:
Merge remote-tracking branch 'origin/master' into concrete-categories

kim-em committed:
merge

urkud committed:
Drop `simp`

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

Commit:
Rename variable

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Rename variable

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Merge remote-tracking branch 'urkud/concrete-categories' into DecLinOrder

urkud committed:
Fix more issues reported by @rwbarton

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

Commit:
Rename variable

Co-Authored-By: Reid Barton <rwbarton@gmail.com>

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

Commit:
Merge branch 'master' into concrete-categories

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

Commit:
merge

urkud committed:
Use `subtype.eq` instead of `subtype.ext.2`

urkud committed:
Cleanup

urkud committed:
Fix compile: now `ring_hom.ext` takes fewer explicit args

urkud committed:
Fix compile

urkud committed:
Run `sanity_check` on all files modified in this branch

* Fix most of the issues reported

* Except for the old ones in `mv_polynomial` (postponed to a separate
  PR) and a false positive in `single_obj`

Commit:
ring_hom.ext changed signature

Commit:
fix signatures

Commit:
more signatures

jcommelin commented on src/algebra/category/CommRing/adjunctions.lean:
```diff
@@ -28,8 +28,8 @@ private def free_obj (α : Type u) : CommRing.{u} := ⟨mv_polynomial α ℤ⟩
   map := λ _ _ f, @ring_hom.of _ _ _ _ (rename f) (rename.is_semiring_hom f),
-  map_id' := λ X, ring_hom.ext _ _ $ funext $ rename_id,
-  map_comp' := λ X Y Z f g, ring_hom.ext _ _ $ funext $ λ p, (rename_rename f g p).symm }
+  map_id' := λ X, ring_hom.ext (funext rename_id),
```
@semorrison Do you know why `tidy` can't take care of these?

kim-em replied:
It can. Yury added the explicit proofs, presumably because this definition is painfully slow to elaborate. (My suspicion is just by the end of `mv_polynomial.lean`, the simp set has grown so unwieldy it's hard to use simp at all.)

urkud replied:
@jcommelin AFAIR, something failed here (probably, `ring_hom.of`), and I wrote everything explicitly.

urkud committed:
Review `∀` vs `Π`

urkud committed:
Remove some unnecessary parentheses

kim-em committed:
add comment

urkud committed:
punit_instances: no need to explicitly provide constants and operations

urkud committed:
Rename `has_forget` to `has_forget₂`

kim-em committed:
add comment, simplify comm_ring punit

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

urkud committed:
Drop `private def free_obj`

kim-em committed:
documentation

kim-em committed:
merge

kim-em committed:
merge

kim-em committed:
merge

kim-em committed:
fixes

jcommelin commented on src/data/mv_polynomial.lean:
```diff
@@ -775,6 +775,8 @@ mv_polynomial.induction_on x
 (λ p n hp, by { rw [eval₂_mul, eval₂_X, hp], exact (is_ring_hom.map_mul f).symm })
 
+/-- Ring homomorphisms out of integer polynomials on a type `σ` are the same as
+functiosn out of the the type `σ`, -/
```
Typo: functions

kim-em committed:
fix typo

kim-em committed:
fix

kim-em committed:
partial fixes, probably all still borked

kim-em committed:
ugh, I mangled the merge

kim-em committed:
fixes

kim-em committed:
comments on uniform space categories

kim-em committed:
cleanup

kim-em committed:
fix botched merge

urkud commented on src/order/category/DecLinOrder.lean:
```diff
@@ -0,0 +1,77 @@
+def of (X : Type u) [preorder X] : Preorder := bundled.of X
+
+instance : unbundled_hom @monotone :=
+⟨@monotone_id, (λ _ _ _ _ _ _ _ _ m₁ m₂, by exactI monotone_comp m₂ m₁)⟩
```
Oh, `monotone_comp` uses the category theory order of arguments. Is it intentional? I mean, most other `*_comp` theorems follow the other convention.

kim-em replied:
I'm not too sure; I just plugged into what existed in the library. @digama0, do you have an opinion here --- should `monotone_comp` have the same order of arguments as the other `*_comp` lemmas?

digama0 replied:
I think things that are called `*_comp` should use the order from function composition. IIRC that's the one that Sebastien implemented across the board a while back.

There are other composition operators called `trans`, for example `equiv.trans`, and these generally match the order of `eq.trans` or `le_trans`, i.e. the >> order. I would use that as the guide.

rwbarton replied:
Yes, let's fix `monotone_comp`, but maybe it is easier to do it in a subsequent PR. Also, we might want to rename it to `monotone.comp` at the same time.

kim-em replied:
Okay, I've noted this as #1468.

kim-em replied:
Okay, I've noted this in #1468.

urkud commented on src/order/category/DecLinOrder.lean:
```diff
@@ -0,0 +1,77 @@
+
+/-- The category of partial orders and monotone maps. -/
+@[reducible]
+def PartialOrder : Type (u+1) := induced_category (bundled.map partial_order.to_preorder.{u})
```
I wonder if having an instance of `unbundled_hom _` has any pros compared to just `concrete_category`. If not, we can completely drop `bundled_hom.induced_category` in favour of this approach.

kim-em replied:
In the last few commits I pursued this, and I think it turns out quite nicely!

urkud replied:
@semorrison It seems that `naturality_bundled` was the only lemma outside of `concrete_category/` formulated for `bundled` instead of `concrete_category`. Thanks for writing `naturality_bundled'`! I copied it to cones, renamed to `naturality_concrete`, and deleted the `_bundled` version. Feel free to amend/fix/revert.

kim-em replied:
We've now gone this route.

urkud commented on src/order/category/DecLinOrder.lean:
```diff
@@ -0,0 +1,77 @@
+
+def of (X : Type u) [linear_order X] : LinOrder := bundled.of X
+
+example : large_category LinOrder.{u} := infer_instance
```
What do you think about changing all these `example`s to `concrete_category`?

kim-em replied:
Yes, good suggestion. I think when I first wrote this I was confused, and thought of `concrete_category` as just an intermediate step to the actual category instance, rather than something strictly better than that category instance.

kim-em replied:
Done.

urkud commented on src/topology/category/TopCommRing.lean:
```diff
@@ -35,8 +38,12 @@ instance : concrete_category TopCommRing.{u} :=
+
+instance (X : TopCommRing) : comm_ring X := X.is_comm_ring
+instance (X : TopCommRing) : topological_space X := X.is_topological_space
+instance (X : TopCommRing) : topological_ring X := X.is_topological_ring
```
What is the difference between `instance (X : TopCommRing) : topological_ring X := X.is_topological_ring` and `attribute [instance] TopCommRing.is_topological_ring`?

kim-em replied:
I think there's a difference: these instances are for, e.g. `comm_ring (X : Type u)`, whereas `TopCommRing.is_comm_ring : comm_ring (X.α)`. These are definitionally but not syntactically equal, so it makes a difference in typeclass search unfortunately.

urkud commented on src/topology/category/UniformSpace.lean:
```diff
@@ -65,6 +65,9 @@ structure CpltSepUniformSpace :=
+-- It should all be provided later by the forgetful functor, but I'm failing to get things working.
+section
 instance : has_coe_to_sort CpltSepUniformSpace :=
 { S := Type u, coe := CpltSepUniformSpace.α }
```
Shouldn't it be a local instance?

kim-em replied:
Yes, it should be, but things break later in the file in a way that I don't understand. If you wanted to have a look that would be very helpful.

kim-em replied:
I managed to clean up `UniformSpace` quite a bit. I'm happier with it, but it's still not as clean as one might hope.

kim-em replied:
Fixed.

jcommelin commented on src/algebra/category/Mon/basic.lean:
```diff
@@ -39,6 +36,9 @@ def of (M : Type u) [monoid M] : Mon := bundled.of M
 ⟨@monoid_hom.to_fun, @monoid_hom.id, @monoid_hom.comp, @monoid_hom.ext⟩
 
+@[to_additive add_monoid]
+instance (x : Mon) : monoid x := x.str
```
```suggestion
instance (M : Mon) : monoid M := M.str
```

jcommelin commented on src/algebra/category/Mon/basic.lean:
```diff
@@ -47,19 +47,19 @@ def CommMon : Type (u+1) := bundled comm_monoid
+Mon.bundled_hom.induced_category @comm_monoid.to_monoid
+
+@[to_additive add_comm_monoid]
+instance (x : CommMon) : comm_monoid x := x.str
```
```suggestion
instance (M : CommMon) : comm_monoid M := M.str
```

jcommelin commented on src/order/category/DecLinOrder.lean:
```diff
@@ -0,0 +1,77 @@
+instance : unbundled_hom @monotone :=
+⟨@monotone_id, (λ _ _ _ _ _ _ _ _ m₁ m₂, by exactI monotone_comp m₂ m₁)⟩
+
+instance (G : Preorder) : preorder G := G.str
```
```suggestion
instance (X : Preorder) : preorder X := X.str
```

jcommelin commented on src/order/category/DecLinOrder.lean:
```diff
@@ -0,0 +1,77 @@
+
+namespace PartialOrder
+
+instance (G : PartialOrder) : partial_order G := G.str
```
```suggestion
instance (X : PartialOrder) : partial_order X := X.str
```

jcommelin commented on src/order/category/DecLinOrder.lean:
```diff
@@ -0,0 +1,77 @@
+
+namespace LinOrder
+
+instance (G : LinOrder) : linear_order G := G.str
```
```suggestion
instance (X : LinOrder) : linear_order X := X.str
```

jcommelin commented on src/order/category/DecLinOrder.lean:
```diff
@@ -0,0 +1,77 @@
+
+namespace DecLinOrder
+
+instance (G : DecLinOrder) : decidable_linear_order G := G.str
```
```suggestion
instance (X : DecLinOrder) : decidable_linear_order X := X.str
```

jcommelin commented on src/topology/category/Top/basic.lean:
```diff
@@ -16,11 +16,11 @@ universe u
 instance concrete_category_continuous : unbundled_hom @continuous :=
 ⟨@continuous_id, @continuous.comp⟩
 
+instance topological_space_unbundled (x : Top) : topological_space x := x.str
```
```suggestion
instance topological_space (x : Top) : topological_space x := x.str
```
This would match the auto-generated name, which is a bit more predictable.

rwbarton replied:
Johan, you got distracted from suggesting to change the variable `x` to `X` :)

jcommelin commented on src/topology/category/TopCommRing.lean:
```diff
@@ -35,8 +38,12 @@ instance : concrete_category TopCommRing.{u} :=
+  forget_faithful := { } }
+end
+
+instance (X : TopCommRing) : comm_ring X := X.is_comm_ring
```
```suggestion
instance (R : TopCommRing) : comm_ring R := R.is_comm_ring
```
Sorry for all these annoying suggestions... but I think eye candy is important PR. And eye-candy is one of the things where the category lib high potential.
(Of course the fact that we have to write these instances is bad... but I think it is non-trivial to automate them away.

jcommelin commented on src/topology/category/TopCommRing.lean:
```diff
@@ -35,8 +38,12 @@ instance : concrete_category TopCommRing.{u} :=
+end
+
+instance (X : TopCommRing) : comm_ring X := X.is_comm_ring
+instance (X : TopCommRing) : topological_space X := X.is_topological_space
```
```suggestion
instance (R : TopCommRing) : topological_space R := R.is_topological_space
```

jcommelin commented on src/topology/category/TopCommRing.lean:
```diff
@@ -35,8 +38,12 @@ instance : concrete_category TopCommRing.{u} :=
+
+instance (X : TopCommRing) : comm_ring X := X.is_comm_ring
+instance (X : TopCommRing) : topological_space X := X.is_topological_space
+instance (X : TopCommRing) : topological_ring X := X.is_topological_ring
```
```suggestion
instance (R : TopCommRing) : topological_ring R := R.is_topological_ring
```

jcommelin commented on src/topology/category/TopCommRing.lean:
```diff
@@ -35,8 +38,12 @@ instance : concrete_category TopCommRing.{u} :=
+instance (X : TopCommRing) : topological_ring X := X.is_topological_ring
 
 /-- Construct a bundled `TopCommRing` from the underlying type and the appropriate typeclasses. -/
 def of (X : Type u) [comm_ring X] [topological_space X] [topological_ring X] : TopCommRing := ⟨X⟩
```
```suggestion
def of (R : Type u) [comm_ring R] [topological_space R] [topological_ring R] : TopCommRing := ⟨R⟩
```

jcommelin reviewed (commented):
A whole bunch of trivial typographic remarks. I really like how streamlined things are getting. Well done @urkud and @semorrison !

kim-em committed:
Update src/algebra/category/Mon/basic.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

kim-em committed:
Update src/order/category/DecLinOrder.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

kim-em committed:
Update src/algebra/category/Mon/basic.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

rwbarton commented on src/category_theory/concrete_category/basic.lean:
```diff
@@ -42,6 +42,9 @@ class concrete_category (C : Type (u+1)) extends category.{u} C :=
 (forget : C ⥤ Type u)
 [forget_faithful : faithful forget]
 
+instance (C : Type (u+1)) [concrete_category C] : has_coe_to_sort C :=
```
This is probably a bad instance, since it matches virtually every type. Can we do without it?

kim-em replied:
Ah. This is a good point, but also makes me sad. :-) We'll have to add a corresponding line to every concrete category.

kim-em replied:
Okay! I think I found a good solution, adding an instance indexed on `induced_category`, that covers nearly all of our use cases so far.

rwbarton commented on src/category_theory/concrete_category/bundled.lean:
```diff
@@ -31,9 +31,6 @@ namespace bundled
 -- Usually explicit instances will provide their own version of this, e.g. `Mon.of` and `Top.of`.
 def of {c : Type u → Type v} (α : Type u) [str : c α] : bundled c := ⟨α, str⟩
 
-instance : has_coe_to_sort (bundled c) :=
```
(By contrast, this instance was fine since it only matches `bundled` things.)

rwbarton commented on src/topology/category/UniformSpace.lean:
```diff
@@ -89,6 +93,7 @@ induced_category.has_forget₂ to_UniformSpace
 end CpltSepUniformSpace
 
 namespace UniformSpace
+local notation `forget` := (forget₂ CpltSepUniformSpace UniformSpace).obj
```
This notation is a bit confusing because there is already something else called `forget` which makes sense in this context. If you just want some short notation, how about `U`?

kim-em replied:
Sounds good.

kim-em committed:
change examples to `concrete_category`

kim-em committed:
Rename types to match mathematical usage

Co-Authored-By: Johan Commelin <johan@commelin.net>

kim-em committed:
Merge branch 'DecLinOrder' of github.com:leanprover-community/mathlib into DecLinOrder

kim-em committed:
fixing cases

kim-em committed:
rename notation to `U`

kim-em committed:
make module doc string

kim-em committed:
solved coercion problem

kim-em committed:
experiments

kim-em committed:
huh, maybe okay

kim-em committed:
cleanup of UniformSpace

kim-em committed:
minor cleanup

kim-em committed:
Merge branch 'master' into DecLinOrder

kim-em committed:
comment

urkud committed:
Add a docstring, reuse an instance

urkud committed:
Formulate naturality of (co)cones over functors to concrete categories without `bundled`

It was here for cocones under the name `naturality_bundled'`. Rename
it `naturality_concrete`, drop the `bundled` version.

urkud committed:
Fix a typo

urkud commented on src/algebra/category/CommRing/adjunctions.lean:
```diff
@@ -41,12 +41,19 @@ def free : Type u ⥤ CommRing.{u} :=
 The free-forgetful adjunction for commutative rings.
 -/
-def adj : free ⊣ forget CommRing :=
+def adj : free ⊣ forget CommRing.{u} :=
```
Do we want to have something like the following (probably, under a different name)? I don't propose to do it in this PR.
```lean
class has_free₂ {C D} [concrete_category C] [concrete_category D] [has_forget₂ C D] :=
(free₂ : D ⥤ C)
(adj : free₂ ⊣ forget₂ C D)

abbreviation has_free {C} [concrete_category C] := has_free₂ C (Type u)
```

kim-em replied:
This sounds great! (For a subsequent PR.)

kim-em committed:
Merge branch 'DecLinOrder' of github.com:leanprover-community/mathlib into DecLinOrder

kim-em committed:
only reducible locally

kim-em committed:
oops, update Group and Mon too

Commit:
cleanup, rfl lemma

kim-em committed:
Merge branch 'master' into DecLinOrder

rwbarton commented on src/category_theory/full_subcategory.lean:
```diff
@@ -26,19 +26,21 @@ section induced
 -/
 
-variables {C : Type u₁} {D : Type u₂} [𝒟 : category.{v} D]
+variables {C : Type u₁} (D : Type u₂) [𝒟 : category.{v} D]
```
I'm surprised you want to make `D` an explicit argument. Shouldn't it normally be inferable from `F`? Are there specific places where this change is needed?

rwbarton commented on src/order/category/DecLinOrder.lean:
```diff
@@ -0,0 +1,87 @@
+open category_theory
+
+/-- The category of preorders and monotone maps. -/
+@[reducible]
```
Are the categories defined in this file intended to be permanently `[reducible]`? I'm not really sure what the ramifications are here, but this PR does go to the trouble of making a bunch of other categories not `[reducible]`...

kim-em replied:
No, this was just an oversight. After the discussion of Zulip I’m now second guessing this whole `local attribute [reducible]` business. In particular I’m worried that it may be contributing to the problems in `CommRing/adjunction.lean`.

rwbarton commented on src/order/category/DecLinOrder.lean:
```diff
@@ -0,0 +1,87 @@
+
+/-- The category of decidable linear orders and monotone maps. -/
+@[reducible]
+def DecLinOrder : Type (u+1) := induced_category LinOrder (bundled.map decidable_linear_order.to_linear_order.{u})
```
Just curious, do you have some specific use for this category in particular?

kim-em replied:
No, I was just looking for a simple chain in the algebraic hierarchy to exercise Yury’s work on concrete categories on. I picked this because it didn’t have many side branches.

rwbarton commented on src/algebra/category/CommRing/basic.lean:
```diff
@@ -25,96 +25,88 @@ universes u v
-  @comm_semiring.to_comm_monoid
-  (λ R₁ R₂ f, f.to_monoid_hom)
-  (by intros; refl)
+has_forget₂.mk'
```
Was there any particular reason to change this one and the next, but not the one above?

rwbarton commented on src/algebra/category/CommRing/adjunctions.lean:
```diff
@@ -41,12 +41,22 @@ def free : Type u ⥤ CommRing.{u} :=
+-- The next two definitions are (unfortunate) "implementation details", helping the elaborator / typeclass search.
+
+/-- Version of `mv_polynomial.hom_equiv` for bundled commutative rings. -/
+def hom_equiv (X : Type u) (R : CommRing) : (mv_polynomial X ℤ →+* R) ≃ (X → R) := hom_equiv R X
```
This clearly indicates that something is wrong. Do you understand the error in `adj` that occurs without this definition?
